java/ql/src/Security/CWE/CWE-400/ReDoS.java

```diff
@@ -0,0 +1,18 @@
+public class ReDoS {
+  private static String EXPONENTIAL_REGEX = "(a+)+";
+
+  public void regexMatchBad(HttpServletRequest request) {
+    // BAD: User provided input is matched against exponential regex using standard
+    // Java regular expression engine
+    java.util.regex.Pattern p = java.util.regex.Pattern.compile(EXPONENTIAL_REGEX);
+    java.util.regex.Matcher m = p.matcher(request.getParameter("input"));
+    boolean matches = m.matches();
+  }
+
+  public void regexMatchGood(HttpServletRequest request) {
+    // GOOD: RE2/J regular expression engine is being used 
+    com.google.re2j.Pattern p = com.google.re2j.Pattern.compile(EXPONENTIAL_REGEX);
+    com.google.re2j.Matcher m = p.matcher(request.getParameter("input"));
+    boolean matches = m.matches();
+  }
+}
```

Comment:
```suggestion
}

```
EOL at EOF

Reply:
I'm not sure about this.
This sample is included in the qhelp. Will this newline at the end look good in the rendered help? Other samples don't have newline at EOF.
@felicitymay, what is your view on this?

Reply:
It's fine with me to change this. But I would like to wait for @felicitymay opinion, because other examples don't have EOL at EOF.

Reply:
My apologies for the delay in replying. I hadn't responded because I was hoping that someone more knowledgeable would answer you. (I'm a content writer, rather than a developer, and work in Windows where EOL and EOF don't really come up.)

While @JLLeitschuh is right to point out that it's normal behavior in linux to end files with an extra newline, I suspect that @ggolawski is right to suggest that for these examples the extra blank line will look strange when they're rendered. @Semmle/java does that sound right?